packages/core/src/components/hotkeys/hotkeysTarget.tsx

```diff
@@ -5,97 +5,87 @@
  */
 
 import * as React from "react";
-import { isFunction, safeInvoke } from "../../common/utils";
+import { IConstructor } from "../../common/constructor";
+import { getDisplayName, isFunction, safeInvoke } from "../../common/utils";
 
 import { IHotkeysProps } from "./hotkeys";
 import { HotkeyScope, HotkeysEvents } from "./hotkeysEvents";
 
 export interface IHotkeysTarget extends React.Component<any, any>, React.ComponentLifecycle<any, any> {
-    /** @internal */
-    globalHotkeysEvents?: HotkeysEvents;
-
-    /** @internal */
-    localHotkeysEvents?: HotkeysEvents;
-
     /**
      * Components decorated with the `HotkeysTarget` decorator must implement
      * this method, and it must return a `Hotkeys` React element.
      */
     renderHotkeys(): React.ReactElement<IHotkeysProps>;
 }
 
-export function HotkeysTarget<T extends { prototype: IHotkeysTarget }>(constructor: T) {
-    const {
-        componentWillMount,
-        componentDidMount,
-        componentWillUnmount,
-        render,
-        renderHotkeys,
-    } = constructor.prototype;
-
-    if (!isFunction(renderHotkeys)) {
-        throw new Error(`@HotkeysTarget-decorated class must implement \`renderHotkeys\`. ${constructor}`);
+export function HotkeysTarget<T extends IConstructor<IHotkeysTarget>>(WrappedComponent: T) {
+    if (!isFunction(WrappedComponent.prototype.renderHotkeys)) {
+        throw new Error(`@HotkeysTarget-decorated class must implement \`renderHotkeys\`. ${WrappedComponent}`);
     }
 
-    // tslint:disable no-invalid-this only-arrow-functions
-    constructor.prototype.componentWillMount = function() {
-        this.localHotkeysEvents = new HotkeysEvents(HotkeyScope.LOCAL);
-        this.globalHotkeysEvents = new HotkeysEvents(HotkeyScope.GLOBAL);
+    return class HotkeysTargetClass extends WrappedComponent {
+        public static displayName = `HotkeysTarget(${getDisplayName(WrappedComponent)})`;
 
-        if (componentWillMount != null) {
-            componentWillMount.call(this);
-        }
-    };
+        /** @internal */
+        public globalHotkeysEvents?: HotkeysEvents;
 
-    constructor.prototype.componentDidMount = function() {
-        // attach global key event listeners
-        document.addEventListener("keydown", this.globalHotkeysEvents.handleKeyDown);
-        document.addEventListener("keyup", this.globalHotkeysEvents.handleKeyUp);
+        /** @internal */
+        public localHotkeysEvents?: HotkeysEvents;
 
-        if (componentDidMount != null) {
-            componentDidMount.call(this);
+        public componentWillMount() {
+            if (super.componentWillMount != null) {
+                super.componentWillMount();
```

Comment:
`safeInvoke`?

Reply:
mmm not a huge fan since you'll have to do `safeInvoke(super.componentWillMount.bind(this))`

Reply:
hmm sure that's a good reason. maybe leave a comment about "preserving `this` binding" ?